python/paddle/distributed/auto_parallel/high_level_api.py

```diff
@@ -0,0 +1,508 @@
+    clear_used_patterns,
+    get_pattern,
+    match_all_patterns,
+    register_used_patterns,
+)
+
+
+class ToDistributedConfig:
+    def __init__(self):
+        self.input_spec = None
+        self.sequence_parallel = False
+
+
+def record_program_ops_pre_hook(layer, inputs):
+    """
+    A pre-hook to mark op numbers before enter layer.forward.
+    """
+    if not in_dygraph_mode():
+        if layer._op_recorder.start < 0:
+            layer._op_recorder.start = len(
+                default_main_program().global_block().ops
+            )
+            # print(f"start program is : {default_main_program()}")
+            layer._op_recorder.is_valid = True
+        else:
+            layer._op_recorder.is_valid = False
+            warnings.warn(
+                f"{layer._full_name} has recorded the op information before. Please check whether you call this layer twice."
+            )
+
+
+def transpose_reshard_embedding_layer_output(layer, inputs, outputs):
+    if hasattr(layer, "current_mesh"):
+        current_mesh = layer.__getattr__("current_mesh")
+        new_output = paddle.transpose(outputs, [1, 0, 2])
+        new_output = dist.reshard(
+            new_output, current_mesh, [dist.Shard(1), dist.Shard(0)]
+        )
+        return new_output
+
+
+def reshard_transpose_attention_layer_input(layer, inputs):
+    # print(f"inputs are {list(inputs)}")
+    # breakpoint()
+    new_inputs = list(inputs)
+    x = new_inputs[0]
+    if hasattr(layer, "current_mesh"):
+        current_mesh = layer.__getattr__("current_mesh")
+        new_x = dist.reshard(x, current_mesh, [dist.Shard(1), dist.Replicate()])
+        new_x = paddle.transpose(new_x, [1, 0, 2])
+        new_inputs[0] = new_x
+        # print(f"new_inputs are: {new_inputs}")
+        # breakpoint()
+        return tuple(new_inputs)
+
+
+def transpose_reshard_attention_layer_output(layer, inputs, outputs):
+    # print(f"outputs are: {outputs}")
+    # breakpoint()
+    attn_out = outputs
+    if hasattr(layer, "current_mesh"):
+        current_mesh = layer.__getattr__("current_mesh")
+        new_attn_out = paddle.transpose(attn_out, [1, 0, 2])
+        new_attn_out = dist.reshard(
+            new_attn_out, current_mesh, [dist.Shard(1), dist.Shard(0)]
+        )
+        # print(f"new_outputs are: {new_attn_out}")
+        # breakpoint()
+        return new_attn_out
+
+
+def reshard_transpose_mlp_layer_input(layer, inputs):
+    new_inputs = list(inputs)
+    mlp_input = new_inputs[0]
+    if hasattr(layer, "current_mesh"):
+        current_mesh = layer.__getattr__("current_mesh")
+        new_mlp_input = dist.reshard(
+            mlp_input, current_mesh, [dist.Shard(1), dist.Replicate()]
+        )
+        # print(f"new_mlp_input is {new_mlp_input}")
+        # breakpoint()
+        new_inputs[0] = new_mlp_input
+        return tuple(new_inputs)
+
+
+def transpose_reshard_mlp_layer_output(layer, inputs, outputs):
```

Comment:
using sequence parallel but not requiring transpose in MLP, is it better to call `reshard_mlp_layer_output`?

Reply:
done